transportable-udfs-annotation-processor/src/main/java/com/linkedin/transport/processor/TransportProcessor.java

```diff
@@ -117,99 +109,81 @@ private void processElements(Set<? extends Element> elements) {
+      TypeMirror iface) {
+    return Streams.concat(Stream.of(typeElement.getSuperclass()), typeElement.getInterfaces().stream())
+        .filter(parent -> _types.isAssignable(parent, iface))
+        .map(parent -> (TypeElement) _types.asElement(parent))
+        .collect(Collectors.toList());
   }
 
   /**
-   * Performs validations on the type hierarchy to verify UDF overloading rules
+   * Finds all elements in the type hierarchy of a {@link TypeElement} which override {@link TopLevelStdUDF} methods
    */
-  private boolean isValidUDFClass(TypeElement udfClassElement) {
-    // Check if any of the interfaces/superclass of the UDF class implement TopLevelStdUDF directly or indirectly
-    TypeMirror superClass = udfClassElement.getSuperclass();
-    boolean superClassImplementsTopLevelStdUDFInterface = _types.isAssignable(superClass, _topLevelStdUDFInterfaceType);
-    LinkedList<TypeMirror> candidateInterfaces = new LinkedList<>();
-    for (TypeMirror iface : udfClassElement.getInterfaces()) {
-      if (_types.isAssignable(iface, _topLevelStdUDFInterfaceType)) {
-        candidateInterfaces.add(iface);
-      }
-    }
-
-    if (superClassImplementsTopLevelStdUDFInterface) {
-      // Rule 1: A UDF class implementing TopLevelStdUDF through a superclass is not supported as of now since it is
-      // hard to determine which interface we want to use as the base for the group of UDF overloading. This is not a
-      // common use case anyway. If required, use composition instead of inheritance.
-      error(Constants.SUPERCLASS_IMPLEMENTS_INTERFACE_ERROR, udfClassElement);
-      return false;
-    } else if (candidateInterfaces.size() > 1) {
-      // Rule 2: More than one interface implementing TopLevelStdUDF can lead to ambiguity with respect to the name of
-      // the UDF and hence is not allowed.
-      error(Constants.MULTIPLE_INTERFACES_ERROR, udfClassElement);
-      return false;
-    } else if (candidateInterfaces.size() == 0) {
-      // Rule 3: A UDF class should implement TopLevelStdUDF either directly or indirectly through an interface only
-      error(Constants.INTERFACE_NOT_IMPLEMENTED_ERROR, udfClassElement);
-      return false;
-    } else if (!_types.isSameType(candidateInterfaces.getFirst(), _topLevelStdUDFInterfaceType)
-        && classOverridesTopLevelStdUDFMethods(udfClassElement)) {
-      // Rule 4: If the class does not directly implement TopLevelStdUDF, then verify that interface methods are not
-      // overriden since it can lead to UDF name ambiguity
-      error(Constants.CLASS_SHOULD_NOT_OVERRIDE_INTERFACE_METHODS_ERROR, udfClassElement);
-      return false;
-    } else {
-      return true;
+  private List<TypeElement> getElementsOverridingTopLevelStdUDFMethods(TypeElement typeElement) {
+    List<TypeElement> elementsOverridingTopLevelStdUDFMethods =
+        getParentsImplementingInterface(typeElement, _topLevelStdUDFInterfaceType).stream()
```

Comment:
Let us discuss naming getParentsImplementingInterface and getElementsOverridingTopLevelStdUDFMethods.

Reply:
Suggestions welcome

Reply:
Looks good after the refactoring. Thank you!